analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,162 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ParameterTypeShouldMatchRouteTypeConstraint : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6800";
+    private const string MessageFormat = "{0}";
+    private const string MessageNoSecondaryLocationFormat = "Parameter type '{0}' does not match route parameter type constraint '{1}' in route '{2}'.";
+    private const string MessageWithSecondaryLocationFormat = "Parameter type '{0}' does not match route parameter type constraint.";
+    private const string SecondaryMessageFormat = "This route parameter has a '{0}' type constraint.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    private static readonly Dictionary<string, string> ConstraintMapping = new(StringComparer.InvariantCultureIgnoreCase)
+    {
+        { "bool", nameof(Boolean) },
+        { "datetime", nameof(DateTime) },
+        { "decimal", nameof(Decimal) },
+        { "double", nameof(Double) },
+        { "float", nameof(Single) },
+        { "guid", nameof(Guid) },
+        { "int", nameof(Int32) },
+        { "long", nameof(Int64) },
+        { "string", nameof(String) }
+    };
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
```

Comment:
This is the second time within 2 weeks, that we should use the unexposed RegisterSymbolStartAction. This kind of analyzer would be much more straightforward with this:
* Register for [SymbolStartAction](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.compilationstartanalysiscontext.registersymbolstartaction?view=roslyn-dotnet-4.6.0) for classes.
* Collect the class route attributes in a list and if not empty, register for PropertyDeclarions.
* In the PropertyDeclaration do a list lookup and raise on mismatch.
 
This is the second example where we should have used such an approach.
https://github.com/SonarSource/sonar-dotnet/pull/8251#discussion_r1371343596